esmvalcore/preprocessor/_regrid.py

```diff
@@ -470,11 +473,19 @@ def regrid(cube, target_grid, scheme, lat_offset=True, lon_offset=True):
     # Return non-regridded cube if horizontal grid is the same.
     if not _horizontal_grid_is_close(cube, target_grid):
 
-        # Perform the horizontal regridding.
+        # Perform the horizontal regridding and preserve data type
+        original_dtype = cube.core_data().dtype
         if _attempt_irregular_regridding(cube, scheme):
             cube = esmpy_regrid(cube, target_grid, scheme)
         else:
             cube = cube.regrid(target_grid, HORIZONTAL_SCHEMES[scheme])
+        try:
+            cube.data = cube.core_data().astype(original_dtype,
+                                                casting='same_kind')
+        except TypeError as exc:
+            logger.warning(
+                "dtype of data changed during regridding from '%s' to '%s': "
+                "%s", original_dtype, cube.core_data().dtype, str(exc))
```

Comment:
I think the potential impact on other data is too large to include it here. Can we keep it separate? Why does it affect ICON particularly?

Reply:
For an unknown (to me) reason the `unstructured_nearest` scheme always converts the data to `float64`, this is why it only affects ICON now. Since I'm using the `try...except` here and `casting='same_kind'` I think we are pretty safe here: The data type is only altered if it doesn't change the data values dramatically (e.g., no `1.5 (float) -> 1 (int)`) or changes from one float type to another.

Reply:
See https://github.com/SciTools/iris/blob/3ff02699b7b6ea341895c4b893febe29670f0435/lib/iris/analysis/trajectory.py#L421 :facepalm: . I'd suggest to report this as an issue to iris and here to do the dtype conversion based on the `scheme`.